cue/cuex/compiler.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2022 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cuex
+
+import (
+	"context"
+	"time"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/build"
+	"cuelang.org/go/cue/cuecontext"
+	"cuelang.org/go/cue/parser"
+	"github.com/spf13/pflag"
+	"k8s.io/klog/v2"
+
+	"github.com/kubevela/pkg/cue/cuex/providers/base64"
+	"github.com/kubevela/pkg/cue/cuex/providers/http"
+	"github.com/kubevela/pkg/cue/cuex/providers/kube"
+	cuexruntime "github.com/kubevela/pkg/cue/cuex/runtime"
+	"github.com/kubevela/pkg/cue/util"
+	"github.com/kubevela/pkg/util/singleton"
+)
+
+const (
+	doKey       = "#do"
+	providerKey = "#provider"
+)
+
+// Compiler for compile cue strings into cue.Value
+type Compiler struct {
+	*cuexruntime.PackageManager
+}
+
+// CompileString compile given cue string into cue.Value
+func (in *Compiler) CompileString(ctx context.Context, src string) (cue.Value, error) {
+	bi := build.NewContext().NewInstance("", nil)
+	bi.Imports = in.PackageManager.GetImports()
+	f, err := parser.ParseFile("-", src)
+	if err != nil {
+		return cue.Value{}, err
+	}
+	if err = bi.AddSyntax(f); err != nil {
+		return cue.Value{}, err
+	}
+	val := cuecontext.New().BuildInstance(bi)
+	return in.Resolve(ctx, val)
+}
+
+// Resolve runs the resolve process by calling provider functions
+func (in *Compiler) Resolve(ctx context.Context, value cue.Value) (cue.Value, error) {
+	newValue := value
+	executed := map[string]bool{}
+	providers := in.PackageManager.GetProviders()
+	for {
+		if ddl, ok := ctx.Deadline(); ok && ddl.After(time.Now()) {
+			return newValue, ResolveTimeoutErr{}
+		}
+		var next *cue.Value
+		// 1. find the next to execute
+		util.Iterate(newValue, func(v cue.Value) (stop bool) {
+			_, done := executed[v.Path().String()]
+			fn, _ := v.LookupPath(cue.ParsePath(doKey)).String()
+			if !done && fn != "" {
+				next = &v
+				return true
+			}
+			return false
+		})
+		if next == nil {
+			break
+		}
+		// 2. execute
+		fn, _ := next.LookupPath(cue.ParsePath(doKey)).String()
```

Comment:
Maybe it's better to add:

```
if !fn.Exists() {
   ...
}
```

Same below since the error is not exactly the same.

Reply:
```
fnV := next.LookupPath(cue.ParsePath(doKey))
if !fnV.Exists() {
   ...
}
fn, _ := fnV.String()
```